### DISCUSSION TITLE
Conversion to CSLA V5.50 Collections

### DISCUSSION THREAD

**Discussion by Art666OTS** [MarimerLLC/csla#2435] [Questions]
I am trying to convert from CSLA 4.11 to CSLA 5.50. The CRUD operations all appear to work fine. 
This is a Business Library using Net Core 5 and CSLA V5.50
Things went pretty smoothly until I looked at the collections. In my previous version I could supply a where clause to the collection and get back the information I was interested in. Below is a comparison of what I had before and what I am attempting to do. The old code is on the top. This would return me a collection of USERS. The only difference is the last line in the using block

```c#
       internal static USERSCollection GetCollection(String filter)
        {
            using (SqlConnection cn = new SqlConnection(DataBase.ApplicationConnection))
            {
                cn.Open();
                using (SqlCommand cm = cn.CreateCommand())
                {
                    cm.CommandType = CommandType.Text;
                    cm.CommandText = String.Format("SELECT * FROM [USERS] WHERE {0}", filter);
                    SqlDataReader dr = cm.ExecuteReader();
                    return new USERSCollection(dr);
                }
            }
        }

       internal static USERSCollection GetCollection(String filter)
        {
            using (SqlConnection cn = new SqlConnection(DataBase.ApplicationConnection))
            {
                cn.Open();
                using (SqlCommand cm = cn.CreateCommand())
                {
                    cm.CommandType = CommandType.Text;
                    cm.CommandText = String.Format("SELECT * FROM [USERS] WHERE {0}", filter);
                    SqlDataReader dr = cm.ExecuteReader();
                    return DataPortal.Create<USERSCollection>(dr);
                }
            }
        }
```

It compiles OK but at run time I get the following.

```
Csla.DataPortalException
  HResult=0x80131500
  Message=DataPortal.Create failed (Csla.BusinessListBase`2[[BusinessLibrary.USERSCollection, BLCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[BusinessLibrary.USERS, BLCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Create](SqlDataReader))
  Source=Csla
  StackTrace:
   at Csla.Reflection.ServiceProviderMethodCaller.FindDataPortalMethod[T](Type targetType, Object[] criteria, Boolean throwOnError)
   at Csla.Server.DataPortal.Create(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.DataPortal`1.Create(Type objectType, Object criteria)
   at Csla.DataPortal`1.Create(Object[] criteria)
   at Csla.DataPortal.Create[T](Object[] criteria)
   at BusinessLibrary.USERSCollection.GetCollection(String filter) in C:\Users\Arthur\Source\Repos\BLCore\BLCore\BO\USERS.cs:line 2042
   at BusinessLibrary.USERSCollection.Collection(String filter) in C:\Users\Arthur\Source\Repos\BLCore\BLCore\BO\USERS.cs:line 2084
   at WebAPI1.Controllers.WeatherForecastController.users() in C:\Users\Arthur\Source\Repos\WebAPI1\WebAPI1\Controllers\WeatherForecastController.cs:line 43
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.SyncObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeActionMethodAsync()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeNextActionFilterAsync()

Inner Exception 1:
TargetParameterCountException: BusinessLibrary.USERSCollection.[Create](SqlDataReader)

Inner Exception 2:
TargetParameterCountException: Csla.BusinessListBase`2[[BusinessLibrary.USERSCollection, BLCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[BusinessLibrary.USERS, BLCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Create](SqlDataReader)

Inner Exception 3:
TargetParameterCountException: Csla.Core.ObservableBindingList`1[[BusinessLibrary.USERS, BLCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Create](SqlDataReader)

Inner Exception 4:
TargetParameterCountException: Csla.Core.MobileObservableCollection`1[[BusinessLibrary.USERS, BLCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Create](SqlDataReader)

Inner Exception 5:
TargetParameterCountException: System.Collections.ObjectModel.ObservableCollection`1[[BusinessLibrary.USERS, BLCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Create](SqlDataReader)

Inner Exception 6:
TargetParameterCountException: System.Collections.ObjectModel.Collection`1[[BusinessLibrary.USERS, BLCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Create](SqlDataReader)

Inner Exception 7:
TargetParameterCountException: System.Object.[Create](SqlDataReader)
```

I cannot seem to find the answer anywhere.

**Comment by rockfordlhotka** [comment]
It looks like the data portal is unable to find a `Create` method that has matching parameters to your `DataPortal.CreateAsync` client-side call.

In other words, you have client-side code like

```c#
  var obj = await DataPortal.CreateAsync<Users>("some filter value");
```

And the data portal expects, in your `Users` class, a data portal operation method like

```c#
  [Create]
  private void Create(string filter)
  {
    var entityData = USERSCollection.GetCollection(filter);
    // load collection from entityData
  }
```

The way the data portal finds server-side operation methods in CSLA 5 is very different (internally) from earlier versions of CSLA, because the data portal now supports dependency injection of parameters. In most cases methods should be backward compatible, but there are edge-cases where issues can happen - especially if you are still using the (soon to be obsolete) `DataPortal_XYZ` naming scheme instead of the new attribute-based scheme.

**Reply by Art666OTS** [reply]
Rockford,

Thank you for your prompt reply. It made me think through what I was doing and I was able to solve the problem by using code I already had and making these small changes.

       internal static USERSCollection GetCollection(String filter)
        {
            using (SqlConnection cn = new SqlConnection(DataBase.ApplicationConnection))
            {
                cn.Open();
                using (SqlCommand cm = cn.CreateCommand())
                {
                    cm.CommandType = CommandType.Text;
                    cm.CommandText = String.Format("SELECT * FROM [USERS] WHERE {0}", filter);
                    SqlDataReader dr = cm.ExecuteReader();
                    USERSCollection UC = new (dr);
                    return UC;
                }
            }
        }
This calls the constructor

        private USERSCollection(SqlDataReader dr)
        {
            MarkAsChild();
            Fetch(dr);
        }
  Which then goes to  Fetch
       private void Fetch(SqlDataReader dr)
        {
            this.RaiseListChangedEvents = false;
            while (dr.Read())
            {
                this.Add(USERS.GetChild(dr));
            }
            this.RaiseListChangedEvents = true;
        }
Now when are in this which I think was the problem because of the missing Create
        internal static USERS GetChild(SqlDataReader dr)
        {
            return DataPortal.Create<USERS>(dr);
        }
This I modified to take the SQLDataReader and from then on it loads the data
        [Create]
        private void Create(SqlDataReader dr)
        {
            Fetch(dr);
        }
        #endregion Factory Methods

I note your comment about "using the (soon to be obsolete) DataPortal_XYZ naming scheme instead of the new attribute-based scheme." I searched for CSLA attribute-based scheme" and didn't get much of interest back. I also scanned though both your V5 books with the same result. I am currently looking through the V4 books up to chapter 2.

I like reading manuals because you always find something you missed before, as soon as I get the update to the attribute scheme done I will be looking at UOW as this looks very interesting.

Thank you very much again.
Arthur

**Reply by hurcane** [reply]
I don't think the attributes are documented in a book yet. The issue https://github.com/MarimerLLC/csla/issues/1102 has the details about the change from the DataPortal_XYZ method names.

**Reply by rockfordlhotka** [reply]
@Art666OTS the best way to keep up with changes is via the release notes in the v5.x and main branches.

* [v5.x branch](https://github.com/MarimerLLC/csla/blob/v5.x/releasenotes.md)
* [main branch](https://github.com/MarimerLLC/csla/blob/main/releasenotes.md)

I absolutely intend on updating the books for CSLA 6, but updating books takes months of effort, and the release notes get updated for every release.

**Reply by Art666OTS** [reply]
Thank you hurcane and Rockford for taking the time to respond to me.

This is just a little bit beyond me at the moment as I have not kept up to date with CSLA. I started with an early V3 and it worked so well I never bothered to change it until Microsoft stop supporting the platform on which it was running.

I then used the "Simple Service implementation with CLSA.NET" by Rockford to upgrade my templates so it would keep working.

Now I think I should upgrade again to CSLA5 and NetCore 5 to keep up with things.

I know it is a big imposition to ask but what I need is a sample/s class files (I have trouble imagining how the fragments that are published all fit together)  that I can use to build the templates for my code generator.

I run purely in ASP.NET and API and like the factory methods and DAL aspects of the Simple service implementation but if there is a newer style I would be happy to look at that.

Thanks
